app/src/main/java/net/squanchy/support/widget/CircleImageView.kt

```diff
@@ -0,0 +1,241 @@
+        return if (isZeroSize) null else bitmapOptional?.let { toNecessaireOptional(it) }
+    }
+
+    private fun toNecessaireOptional(bitmap: Bitmap): Necessaire {
+        val bounds = calculateBounds()
+        val paint = createPaintFor(bitmap, bounds)
+
+        val radius = Math.min(bounds.height() / 2.0f, bounds.width() / 2.0f)
+
+        return Necessaire(
+                bounds = bounds,
+                paint = paint,
+                radius = radius
+        )
+    }
+
+    private fun createPaintFor(bitmap: Bitmap, bounds: RectF): Paint {
+        val shader = BitmapShader(bitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
+        val bitmapWidth = bitmap.height
+        val bitmapHeight = bitmap.width
+        setTransformation(shader, bounds, bitmapWidth, bitmapHeight)
+
+        val paint = Paint()
+        paint.isAntiAlias = true
+        paint.shader = shader
+        return paint
+    }
+
+    private fun calculateBounds(): RectF {
+        val availableWidth = width - paddingLeft - paddingRight
+        val availableHeight = height - paddingTop - paddingBottom
+
+        val sideLength = min(availableWidth, availableHeight)
+
+        val left = paddingLeft + (availableWidth - sideLength) / 2f
+        val top = paddingTop + (availableHeight - sideLength) / 2f
+
+        return RectF(left, top, left + sideLength, top + sideLength)
+    }
+
+    private fun setTransformation(shader: Shader, bounds: RectF, bitmapWidth: Int, bitmapHeight: Int) {
+        val scale: Float
+        val translationX: Float
+        val translationY: Float
+
+        if (heightDictatesScale(bounds, bitmapWidth, bitmapHeight)) {
+            scale = bounds.height() / bitmapHeight.toFloat()
+            translationX = (bounds.width() - bitmapWidth * scale) / 2f
+            translationY = 0f
+        } else {
+            scale = bounds.width() / bitmapWidth.toFloat()
+            translationX = 0f
+            translationY = (bounds.height() - bitmapHeight * scale) / 2f
+        }
+
+        val matrix = createMatrixFor(scale, bounds, translationX, translationY)
+        shader.setLocalMatrix(matrix)
+    }
+
+    private fun heightDictatesScale(bounds: RectF, bitmapWidth: Int, bitmapHeight: Int): Boolean {
+        return bitmapWidth * bounds.height() > bounds.width() * bitmapHeight
+    }
+
+    private fun createMatrixFor(scale: Float, bounds: RectF, translationX: Float, translationY: Float): Matrix {
+        val matrix = Matrix()
+        matrix.setScale(scale, scale)
+        val x = bounds.left + intCeil(translationX)
+        val y = bounds.top + intCeil(translationY)
+        matrix.postTranslate(x, y)
+        return matrix
+    }
+
+    private fun intCeil(value: Float): Int = (value + ROUNDING_UP_FLOAT).toInt()
+
+    override fun onDraw(canvas: Canvas) {
+        super.onDraw(canvas)
+
+        necessaireOptional?.also { draw(canvas, it) }
+    }
+
+    private fun draw(canvas: Canvas, necessaire: Necessaire) {
```

Comment:
Could be inlined, it's a one-liner whose `with` is acting the same as the `also` on line 208